Enable etcd and safekeepers in all tests

SomeoneToIgnore opened the pull request:
Part of https://github.com/neondatabase/neon/pull/1619

This is a PR that should create preliminary breakage before the main breakage is implemented.
* starts one safekeeper for every test 
I'm not sure if it uses that though, I should I also remove [this](https://github.com/neondatabase/neon/blob/ef40e404cf15cc335fbf6a226879e5358aa628eb/test_runner/fixtures/zenith_fixtures.py#L1543) line from fixtures? Anywhere else?
* starts an etcd for every test

Makes both pageserver and safekeeper binaries to require etcd broker endpoints specified for them in configs/startup args.

Without the endpoints, no etcd connection is possible.
Without etcd connection, no WAL streaming from Safekeeper to Pageserver is possible after the WIP PR mentioned above.

SomeoneToIgnore commented:
At least the e2e tests fail expectedly correctly.
I'll adjust those later, but would be glad to know if I'm doing everything right and no direct compute streaming tests are left.

SomeoneToIgnore commented:
Console PR: https://github.com/neondatabase/cloud/pull/1195

SomeoneToIgnore commented:
Just realized that I have to adjust neon_base cli now to start etcd on init and stop at least.

Another good topic to discuss is whether to make the broker_endpoints mandatory as it is now: their absence in the new model makes both binary pretty useless, but for local hacking at least, it might be ok to start without the broker?

arssher commented on neon_local/src/main.rs:
```diff
@@ -896,20 +896,23 @@ fn handle_safekeeper(sub_match: &ArgMatches, env: &local_env::LocalEnv) -> Resul
 
-fn handle_start_all(sub_match: &ArgMatches, env: &local_env::LocalEnv) -> Result<()> {
+fn handle_start_all(sub_match: &ArgMatches, env: &local_env::LocalEnv) -> anyhow::Result<()> {
+    etcd::start_etcd_process(env)?;
```
In current coding etcd is started by both python code and neon_local, we clearly need one of them. The question is which one. Python code gives slightly more handy manual control, but leaving neon_local untested is not nice. I think we can do everything from neon_local, but make it more functional:

- on stop wait handle, bail out if stop failed;
- wait for start by checking status;
- make the whole etcd managing business optional (it might be running elsewhere) -- e.g. bool flag in conf;
- probably add separate etcd start/stop commands. For example, currently ZenithEnvBuilder on __exit__ doesn't call `neon stop`, it shuts down services individually. etcd start/stop can also be useful for some tests.

Or just sit with python version for some more time.

SomeoneToIgnore replied:
I've decided to sit with the python for now without extra big changes.
More on the topic: https://github.com/neondatabase/neon/pull/1648#issuecomment-1125948912

arssher commented:
> Another good topic to discuss is whether to make the broker_endpoints mandatory as it is now: their absence in the new model makes both binary pretty useless, but for local hacking at least, it might be ok to start without the broker?

Yes, I would keep it optional. For example, for benchmarking safekeepers it is useful to have working setup without pageserver at all (we actually have such test).

arssher commented:
> I'm not sure if it uses that though, I should I also remove [this](https://github.com/neondatabase/neon/blob/ef40e404cf15cc335fbf6a226879e5358aa628eb/test_runner/fixtures/zenith_fixtures.py#L1543) line from fixtures? Anywhere else?

Yes, with this PR all tests use safekeeper. As for this line -- its fate should be the same as `callmemaybe_connstring` configuration in `setup_pg_conf`. If we don't support setup without safekeepers at all, it can be removed. But I'm dubious on that. Such setup is still might be useful for tests/experiments/local deployment. But for this to work, we must either don't drop callmemaybe support from pageserver (which I don't expect to be hard actually) or reverse the flow (initiate replication connection by compute), which is a good idea already discussed, but more involved.

arssher commented on control_plane/src/local_env.rs:
```diff
@@ -59,13 +59,7 @@ pub struct LocalEnv {
-    /// A prefix to all to any key when pushing/polling etcd from a node.
-    #[serde(default)]
-    pub broker_etcd_prefix: Option<String>,
+    pub etcd_broker: EtcdBroker,
```
Should also add this to `simple.conf`

SomeoneToIgnore replied:
That resulted in https://github.com/neondatabase/neon/pull/1697 eventually

petuhovskiy reviewed (commented):
Looks fine, but I have several questions:
- What's with removing etcd from python fixtures? Can we handle it all in control_plane?
- What's current approach for etcd reuse? Are we launching separate etcd instances for each test? Can we have one for all the tests?

SomeoneToIgnore commented on control_plane/src/local_env.rs:
```diff
@@ -412,3 +457,19 @@ fn base_path() -> PathBuf {
     }
 }
+
+// TODO kb split fs checks and parsing methods, to test the conf for completeness
```
We should, sorry for the amount of pushes, I was fighting with the bash scripting to make our `broker_endpoints` propagated to every e2e test.
Luckily, this was the last one and tests are green now: https://github.com/neondatabase/cloud/actions/runs/2316232005

This means, we've got all possible places `broker_endpoints` covered and I can throw away the code that fails on this config value absence and address the rest of the comments.

To justify the extra pushes here, I've been addressing some comments and this TODO is a result of the `simple.conf` comment: I want to create a test for parsing the file and left a reminder.

SomeoneToIgnore commented:
> What's with removing etcd from python fixtures? Can we handle it all in control_plane?

We can and we should, but I don't think I want to do this now, since it's quite a big change per se.
I'm into the idea fully though, we can create a follow-up task for that.

> What's current approach for etcd reuse? Are we launching separate etcd instances for each test? Can we have one for all the tests?

So far, it's a single cluster of a single etcd node in e2e tests: https://github.com/neondatabase/cloud/pull/1241/files#diff-e45e45baeda1c1e73482975a664062aa56f20c03dd9d64a827aba57775bed0d3
and a separate etcd per test here, with its own logs directory.
Looks like an overkill indeed, I'll move it out same way we have the remote storage now when addressing the rest of the unresolved comments.

SomeoneToIgnore commented on control_plane/simple.conf:
```diff
@@ -9,3 +9,6 @@ auth_type = 'Trust'
 http_port = 7676
+
+[etcd_broker]
+broker_endpoints = ['127.0.0.1::2379']
```
There were no tests for that, I've added those separately: https://github.com/neondatabase/neon/pull/1697

Funny that `['127.0.0.1::2379']` gets parsed as `Url` just fine:

```
broker_endpoints: [
            Url {
                scheme: "localhost",
                cannot_be_a_base: true,
                username: "",
                password: None,
                host: None,
                port: None,
                path: ":2379",
                query: None,
                fragment: None,
            },
        ],
```

Anyway, let's continue there.

SomeoneToIgnore commented on test_runner/fixtures/zenith_fixtures.py:
```diff
@@ -1891,7 +1896,13 @@ def test_output_dir(request: Any) -> str:
+                       'pg_subtrans',
+                       'pg_logical',
+                       'pg_replslot/wal_proposer_slot',
+                       'pg_xact'))
```
Reverted back, that was wrong, thanks.

SomeoneToIgnore commented:
Ok, had addressed all feedback, but please ask if anything's unclear, can be reviewed further.

Few action points: 
* there's now one global etcd and mock s3 servers in our python tests (s3 is still enabled per test, so test logic is still the same)
* a separate simple.conf set of fixes: https://github.com/neondatabase/neon/pull/1697
* all `broker_endpoints` vars were first made mandatory and a cloud PR was created to ensure that e2e tests pass and have endpoints: https://github.com/neondatabase/cloud/pull/1241
Then, the mandatory requirement was reverted, so now pageserver in this PR does nothing about the variables besides parsing, safekeeper issues a warning on empty endpoints

LizardWizzard commented:
Are we really sure that we want to start new etcd instance on per-test basis? IMO the better option is to introduce installation id in etcd paths as @arssher suggested and use it to namespace different test envs in a single etcd instance. 

Per test etcd set up will be extremely resource demanding for parallel test runs.

petuhovskiy commented:
I looked at the CI, the only failed job was `pg_regress-tests-debug` that failed with `RuntimeError: timed out waiting 294s for etcd start: HTTPConnectionPool(host='127.0.0.1', port=16001)`. Does it mean that this PR is almost ready now?

IMO we can merge this PR once all tests in CI are stable and fix other things later (better to create an issue for that).

SomeoneToIgnore commented:
Prefixes is a nice idea actually, had not implemented that, but considered the PR to be ready indeed, wanted to create a separate issue for better python-control_plane separation and had thought that the rest of the feedback got addressed.

I'll run the tests a few more times, to see what's wrong with the startups.

arssher reviewed (approved):
Perhaps remove `start_etcd_process` from `handle_start_all` if for now we use python version in tests? Fixtures don't use `start all`, but that's confusing on the first look.

SomeoneToIgnore commented:
I've managed to overcome the flackines via [somewhat obscure way](https://app.circleci.com/pipelines/github/neondatabase/neon/6406/workflows/62dcba74-8562-4744-b1ea-b2f994304e6b/jobs/64411/artifacts):
<img width="540" alt="image" src="https://user-images.githubusercontent.com/2690773/168681811-625e2433-a864-4ac2-864f-ffdbd30615ee.png">

Apparently, the `@pytest.fixture(scope='session')` dependencies are different for parallel tests, so they start different etcds.
I could try and write even more code to detect an already running etcd, but decided to keep things easier for now and keep this set up.
If a few reruns won't show failures, I'll merge the PR.

SomeoneToIgnore commented:
Follow-up tasks:

* https://github.com/neondatabase/neon/issues/1716
* https://github.com/neondatabase/neon/issues/1717

I seem to have addressed everything by this point?
